### PR TITLE
Support for ConvLSTM2D

### PR DESCRIPTION
**Pull request by dikatok**
Support for ConvLSTM2D.

Fixes https://github.com/tensorflow/tfjs/issues/3637.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/3702)

### CODE REVIEW

**Comment by dikatok** [comment]
Origin issue https://github.com/tensorflow/tfjs/issues/3637

**Comment by dikatok** [comment]
@rthadur is there anyone else to notify about this new feature?

**Comment by dikatok** [comment]
@dsmilkov @pyu10055 hi guys, is it possible to run specific unit tests?

File: tfjs-layers/src/initializers.ts [hunk 1]
```diff
@@ -545,9 +545,8 @@ export class Orthogonal extends Initializer {
 
   apply(shape: Shape, dtype?: DataType): Tensor {
     return tidy(() => {
-      if (shape.length !== 2) {
-        throw new NotImplementedError(
-            'The Orthogonal Initializer does not support non-2D shapes yet.');
+      if (shape.length < 2) {
```
Comment (dikatok) [inline tfjs-layers/src/initializers.ts:548]: I modified this check since the default recurrent initializer in ConvLSTM2DCell is orthogonal which needs to support 4D. 

This update seems to be inline too with Python implementation (link below). 

Currently, I have only run unit tests in layer repo. Do you have any concerns?

https://github.com/tensorflow/tensorflow/blob/6016e8009ff8d9b3afa5ae5bafd052ed51d22f74/tensorflow/python/ops/init_ops.py#L571

**Comment by dikatok** [comment]
@dsmilkov @pyu10055 I encountered a compatibility issue between BatchNorm and ConvLSTM2D where the BatchNorm does not support tensor rank past 4. 

Is this limited rank support kernel implementation related?

**Comment by dikatok** [comment]
@pyu10055 would you mind taking a look at this? 

I didn't expect this to be so large 😭

**Comment by dikatok** [comment]
@pyu10055 ping

**Comment by caisq** [comment]
@dikatok Sorry for the delay. I'm not reviewing this PR and will get back to you with a review by end of day today.

**Comment by dikatok** [comment]
@caisq 

```
class ConvRNN2D extends RNN {
```
How verbose would this docstring need to be?
<br /><br />
```
if (this.cell.dropoutMask != null) {
        tfc.dispose(this.cell.dropoutMask);
        this.cell.dropoutMask = null;
      }
```
Actually, I was following the way it is structured in recurrent cells, not quite sure about the why either since I could not find any from the commit changes.
<br /><br />
```
generateDropoutMask
```
Should I update the one in recurrent.ts with mine? I think it looks cleaner.

**Comment by dikatok** [comment]
@caisq Done!

Thanks for the quick feedback loop, by the way, I thought the PR would be dangling for quite a while.

Also, would you mind taking a look at my previous comments up there and see if you have any suggestions?

**Comment by caisq** [comment]
Looks like there are lint errors. 

> ERROR: (no-unnecessary-type-assertion) /workspace/tfjs-layers/src/layers/convolutional_recurrent.ts[160, 35]: This assertion is unnecessary since it does not change the type of the expression.

@dikatok Can you run `yarn lint` locally and fix all the errors?

**Comment by caisq** [comment]
@dikatok The code snippet in the comments should use `tf.*` instead of `tfc.*`. This is why the CI is failing.

```js
const filters = 3;
const kernelSize = 3;

const sequenceLength = 1;
const size = 5;
const channels = 3;

const inputShape = [sequenceLength, size, size, channels];
const input = tfc.ones(inputShape);

const cell = tf.layers.convLstm2dCell({filters, kernelSize});

cell.build(input.shape);

const outputSize = size - kernelSize + 1;
const outShape = [sequenceLength, outputSize, outputSize, filters];

const initialH = tfc.zeros(outShape);
const initialC = tfc.zeros(outShape);

const [o, h, c] = cell.call([x, initialH, initialC], {});
```

**Comment by dikatok** [comment]
@caisq ah okay, will fix it later

**Comment by dikatok** [comment]
@caisq can this jsdoc assertion performed locally? or it is only run in CI?

**Comment by caisq** [comment]
@dikatok The snippet-testing code is here:
https://github.com/tensorflow/tfjs/blob/master/tfjs-layers/scripts/test_snippets.ts

I'm not sure about the correct command to invoke it. But you can try figuring it out or maybe @pyu10055 can tell you more. It runs on CI for sure.

**Comment by caisq** [comment]
Actually, running `yarn test-snippet` in the tfjs-layers folder should be the way. See:
https://github.com/tensorflow/tfjs/blob/master/tfjs-layers/package.json#L65

**Comment by dikatok** [comment]
Ah my bad, it should be fixed now.

About these issues I encountered earlier, should they be handled in separate issues/PR in the future?

- batchNormalization if returnSequences: true Error: batchNormalization is not implemented for array of rank 5 yet
- running ConvLSTM2DCell via RNN or StackedRNN, although this should be discouraged
- goBackwards: true Error: WebGL backend: Reverse of rank-5 tensor is not yet supported
- gather (webgl) Error: Gather for rank 5 is not yet supported

**Comment by caisq** [comment]
@pyu10055 PTAL? I think PR is ready to merge.

**Comment by dikatok** [comment]
@caisq BTW, is it acceptable for outside contributors like me to do code refactoring? I kind of want to refactor the `recurrent.ts` file to make it easier to digest. Probably will start with arguments inheritance and the unnecessary getters.

**Comment by caisq** [comment]
@dikatok I think it is acceptable. I'd suggest doing it as a separate PR from this one, though.
